### PR TITLE
improved version of HoughCircles (HOUGH_GRADIENT_ALT method)

### LINKED ISSUE #16187: HoughCircles accuracy decreased in opencv 4.1.1 after upgrade from 3.1.0
##### System information (version)

- OpenCV => 4.1.1
- Operating System / Platform => Windows 64 Bit
- Compiler =>Visual Studio 2019

##### Detailed description

The HoughCircles() from version 3.1.0 was used to detect the circles in 8bit gray value images with a resolution of 320x240 pixel. After an upgrade to version 4.1.1 the HoughCircles() detection accuraccy - the euclidean distance between ground truth (HoughCircles 3.1.0) and a detected centre decreased of 10% of an image resolution. Similar accuraccy loss has been observed by the radius values. 

dist_centers_graph
Circle position error in pixels tested on 438 images

diff_radii_graph
Radius values error in pixels tested on 438 images
histogram_centre_dist
Histogram with frequencies of error classes.

##### Steps to reproduce

1. A sample with HoughCircles call has been compiled with an opencv 3.1.0 build from source code, the results and parameters have been saved as metadata.
2. A sample with HoughCircles call has been compiled with an opencv 4.1.1 build from source code, the parameters and results from 3.1.0 have been read. The difference has been saved as metadata.
3. Results have been visualized.

###### Sample source code
````
#include <iostream>

#include <opencv2/core.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc.hpp>

#include <string>
#include <iostream>
#include <fstream>
#include <filesystem>
#include <list>


using namespace cv;
using namespace std;

namespace fs = std::filesystem;

bool read_params(const fs::path& prm_file_path, int& rnX, int& rnY, int& rnW, int& rnH, double& rdMinDist, double& rdCanny, double& rdNumVotes, int& rnMinRad, int& rnMaxRad, int& rnNumCircles, double& rdCentX, double& rdCentY, double& rdRadius)
{
    if (!fs::exists(prm_file_path))
    {
        return false;
    }

    ifstream oStream;

    oStream.open(prm_file_path.string());

    if (!oStream.is_open())
    {
        return false;
    }

    int nTmp;

    oStream >> nTmp;                // cnScale
    oStream >> rnX;                 // X
    oStream >> rnY;                 // Y
    oStream >> rnW;                 // W
    oStream >> rnH;                 // H
    oStream >> rdMinDist;           // min dist
    oStream >> rdCanny;             // canny thresh
    oStream >> rdNumVotes;          // num votes
    oStream >> rnMinRad;            // min radius
    oStream >> rnMaxRad;            // max radius
    oStream >> rnNumCircles;        // number of circles found
    if (0 != rnNumCircles)
    {
        oStream >> rdCentX;             // first circle center X
        oStream >> rdCentY;             // first circle center Y
        oStream >> rdRadius;            // first circle radius
    }
    else
    {
        rdCentX = 0.0;
        rdCentY = 0.0;
        rdRadius = 0.0;
    }

    oStream.close();

    return true;
}

bool find_hough_circles(const cv::Mat& input_img, const int& nX, const int& nY, const int& nW, const int& nH, const double& dMinDist, const double& dCanny, const double& dNumVotes, const int& nMinRad, const int& nMaxRad, int& rnNumCircles, double& rdCentX, double& rdCentY, double& rdRadius)
{
    Rect roi = Rect(nX, nY, nW, nH);
    Mat roi_image = input_img(roi);

    vector<Vec3f> aCircles;

    try
    {
        HoughCircles(roi_image, aCircles, cv::HOUGH_GRADIENT, 2.0, dMinDist, dCanny, dNumVotes, nMinRad, nMaxRad);
    }
    catch (const cv::Exception & e)
    {
        std::cout << "Exception thrown: " << e.what() << std::endl;
        return false;
    }

    rnNumCircles = (int)aCircles.size();

    if (rnNumCircles)
    {
        rdCentX = aCircles[0][0];
        rdCentY = aCircles[0][1];
        rdRadius = aCircles[0][2];
    }
    else
    {
        rdCentX = 0.0;
        rdCentY = 0.0;
        rdRadius = 0.0;
    }

    return true;
}

fs::path find_dbg_filename(const list<fs::path>& dbg_names, const string& dbg_subname)
{
    for (const auto& entry : dbg_names)
    {
        const fs::path& entry_stem = entry.stem();

        if (entry_stem.string() == dbg_subname)
        {
            return entry;
        }
    }

    return fs::path();
}

int main(int argc, char* argv[])
{
    std::cout << "Starting test with opencv 4.1.1\n";

    if (3 == argc)
    {
        if (0 == strcmp("-d", argv[1]))
        {
            std::cout << "Starting the test in directory: " << argv[2] << std::endl;
            string dir_path(argv[2]);

            if (!fs::exists(dir_path) ||
                !fs::is_directory(dir_path))
            {
                std::cout << "Source directory doesn't exist or given path is not a directory. The test will terminate..." << std::endl;
                return -1;
            }

            if (0 == strcmp(&dir_path.back(), "/") ||
                0 == strcmp(&dir_path.back(), "\\"))
            {
                dir_path.pop_back();
            }

            ofstream result_file;
            fs::path src_dir_path(dir_path);
            fs::path result_file_path(fs::current_path());
            fs::path result_file_stem = src_dir_path.stem();
            result_file_path += "\\results_";
            result_file_path += result_file_stem;
            result_file_path += "_411.txt";

            //result_file.open(dir_path + "\\" + "results.txt", ofstream::out | ofstream::trunc);
            result_file.open(result_file_path, ofstream::out | ofstream::trunc);

            if (!result_file.is_open())
            {
                cout << "Can't open result file!" << endl;
                return -1;
            }

            result_file << "Image name" << "\t" << "Num circles found (difference)" << "\t" << "first found circle X (difference)" << "\t"
                << "first found circle Y (difference)" << "\t" << "first found circle R (difference)" << "\t" << "Num circles found (orig)" << "\t" << "first found circle X (orig)" << "\t"
                << "first found circle Y (orig)" << "\t" << "first found circle R (orig)" << "\t" << "Num circles found" << "\t" << "first found circle X" << "\t" << "first found circle Y" << "\t"
                << "first found circle R" << "\t" << "ROI X" << "\t" << "ROI Y" << "\t" << "ROI W" << "\t" << "ROI H" << "\t" << "HC Param min dist" << "\t"
                << "HC Param Canny Threashold" << "\t" << "HC Param Min Votes" << "\t" << "HC Param min r" << "\t" << "HC Param max r" << endl;

            list <fs::path> txt_entries;
            list <fs::path> img_in_entries;
            list <fs::path> img_dbg_entries;

            for (const auto& entry : fs::directory_iterator(dir_path))
            {
                string extension = entry.path().extension().string();

                if (entry.is_regular_file())
                {
                    if (".txt" == entry.path().extension() &&
                        "results" != entry.path().stem())
                    {
                        txt_entries.push_back(entry);
                    }
                    else if (".png" == extension ||
                        ".jpg" == extension ||
                        ".bmp" == extension)
                    {
                        if (string::npos != entry.path().string().find("_dbg"))
                        {
                            img_dbg_entries.push_back(entry);
                        }
                        else
                        {
                            img_in_entries.push_back(entry);
                        }
                    }
                    else
                    {
                        // nothing to do here (unknown file)
                    }
                }
            }

            //search the circle
            for (const auto& txt_entry : txt_entries)
            {
                // test
                double dDist, dCanny, dVotes, dCx, dCy, dR, dCxOrig, dCyOrig, dROrig;
                int nX, nY, nW, nH, nMin, nMax, nCount, nCountOrig;
                if (read_params(txt_entry, nX, nY, nW, nH, dDist, dCanny, dVotes, nMin, nMax, nCountOrig, dCxOrig, dCyOrig, dROrig))
                {
                    string dbg_img_path_name = txt_entry.stem().string();
                    dbg_img_path_name += "_dbg";
                    fs::path dbg_img_path = find_dbg_filename(img_dbg_entries, dbg_img_path_name);

                    Mat input_img;
                    input_img = imread(dbg_img_path.string(), 0);

                    if (input_img.data)
                    {
                        flip(input_img, input_img, 0);

                        if (find_hough_circles(input_img, nX, nY, nW, nH, dDist, dCanny, dVotes, nMin, nMax, nCount, dCx, dCy, dR))
                        {
                            int nDCount = abs(nCountOrig - nCount);
                            float dDX = abs(dCxOrig - dCx);
                            float dDY = abs(dCyOrig - dCy);
                            float dDR = abs(dROrig - dR);

                            result_file << txt_entry.stem().string() << "\t" << nDCount << "\t" << std::fixed << dDX << "\t" << dDY << "\t" << dDR << "\t" << nCountOrig << "\t" << dCxOrig << "\t"
                                << dCyOrig << "\t" << dROrig << "\t" << nCount << "\t" << dCx << "\t" << dCy << "\t"
                                << dR << "\t" << nX << "\t" << nY << "\t" << nW << "\t" << nH << "\t" << dDist << "\t"
                                << dCanny << "\t" << dVotes << "\t" << nMin << "\t" << nMax << endl;

                            if (nCount)
                            {
                                try
                                {
                                    Mat out_img;
                                    cvtColor(input_img, out_img, cv::COLOR_GRAY2RGB);
                                    Rect roi = Rect(nX, nY, nW, nH);
                                    Mat roi_image = out_img(roi);
                                    circle(roi_image, cv::Point(dCx, dCy), (int)dR, cv::Scalar(255, 255, 0));
                                    fs::path out_img_path(src_dir_path);
                                    out_img_path += "\\";
                                    out_img_path += txt_entry.stem().string();
                                    out_img_path += "_out_411.png";
                                    imwrite(out_img_path.string(), roi_image);
                                }
                                catch (const cv::Exception & e)
                                {
                                    cout << e.what() << endl;
                                }
                            }
                        }
                        else
                        {
                            cout << "Error occured while searchign hough circles in: " << dbg_img_path_name << endl;
                            result_file << "Error occured while searchign hough circles in: " << dbg_img_path_name << endl;
                        }
                    }
                    else
                    {
                        cout << "Can't read image: " << dbg_img_path_name << endl;
                        result_file << "Can't read image: " << dbg_img_path_name << endl;
                    }
                }
                else
                {
                    cout << "Can't read parameter file: " << txt_entry << endl;
                    result_file << "Can't read parameter file: " << txt_entry << endl;
                }
            }

            std::cout << "Results written to: " << result_file_path << std::endl;
        }
        else
        {
            std::cout << "Wrong number of parameters, test will terminate..." << std::endl;
            return -1;
        }
    }
    else
    {
        std::cout << "Wrong number of parameters, test will terminate..." << std::endl;
        return -1;
    }

    return 0;
}
````
Following test data show the input images (_dbg) results of the 3.1.0 (magenta circles) and results of the 4.1.1 (cyan circles) The tendence of the algorithm to provide mostly large circles can be in the 4.1.1 suppressed by the normalisation (votes/diameter) as the new algorithm provides also the votes of the found circles. This leads to not much better results. The text files contain the parameters which have been used during circles matching. The data have to be put in the directory and path to the directory has to be used as a parameter in the command line.

04925_01_013_04925d080_green_out
04925_01_013_04925d080_green_out_411
04936_01_021_04936d147_brown_dbg
04936_01_021_04936d147_brown_out
04936_01_021_04936d147_brown_out_411
[04890_02_021_04890d462_brown.txt](https://github.com/opencv/opencv/files/3977776/04890_02_021_04890d462_brown.txt)
04890_02_021_04890d462_brown_dbg
04890_02_021_04890d462_brown_out
04890_02_021_04890d462_brown_out_411
04908_01_001_04908d098_hazel_dbg
04908_01_001_04908d098_hazel_out
04908_01_001_04908d098_hazel_out_411
04913_01_001_04913d010_blue_dbg
04913_01_001_04913d010_blue_out
04913_01_001_04913d010_blue_out_411
04925_01_013_04925d080_green_dbg

### PR DESCRIPTION
**Pull request by vpisarev**
relates #16187 (on one image the eye is not detected with param2=0.9, need to lower it down to 0.85). Overall, the quality is much better than the existing HOUGH_GRADIENT method.

Usage:

```
Mat smoothed;
GaussianBlur(grayscale_image, smoothed, Size(7, 7), 1.5, 1.5);
HoughModes method = HOUGH_GRADIENT_ALT;
double dp = 1.5;
double minDist = 10;
double cannyThreshold = 300;
double minCos2 = 0.9; // 0.85 for one of the eye images
int minRadius = 7;
int maxRadius = 500;
std::vector<Vec3f> circles;
HoughCircles(smoothed, circles, method, dp, minDist,
           cannyThreshold, minCos2, minRadius, maxRadius);
```

this code detects circles on most of the images that I have (~100 images; the proper regression test will be added later). When it does not, try to narrow down the radius search range `[minRadius, maxRadius]` and decrease a bit `minCos2` from 0.9 down to 0.85..0.75.

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] (need to add more tests in the future) There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/imgproc/src/hough.cpp [hunk 1]
```diff
@@ -1710,6 +1711,467 @@ static void HoughCirclesGradient(InputArray _image, OutputArray _circles,
...
+
+    for( int x = 0; x < mask.cols; x++ ) mdata[x] = mdata[(mask.rows-1)*mstep + x] = (uchar)1;
+    for( int y = 0; y < mask.rows; y++ ) mdata[y*mstep] = mdata[y*mstep + mask.cols-1] = (uchar)1;
+    mdata += mstep + 1;
+
+    for( int y = 0; y < edges.rows; y++ )
+    {
+        for( int x = 0; x < edges.cols; x++ )
+        {
+            if(!edgeData[y*estep + x] || mdata[y*mstep + x])
```
[the remark this answers is not in the document]
Comment (vpisarev) [reply modules/imgproc/src/hough.cpp]: This is the inner loop where performance matters. Address arithmetic is important here and will not be replaced with .at

File: modules/imgproc/src/hough.cpp [hunk 2]
```diff
@@ -1710,6 +1711,467 @@ static void HoughCirclesGradient(InputArray _image, OutputArray _circles,
...
+    const short* dxData = Dx.ptr<short>();
+    const short* dyData = Dy.ptr<short>();
+    int dxystep = (int)Dx.step1();
+    uchar* mdata = mask.ptr<uchar>();
+    int mstep = (int)mask.step1();
+
+    circles.clear();
+    std::vector<Vec4f> nz;
+
+    std::vector<Point> stack;
```
[the remark this answers is not in the document]
Comment (vpisarev) [reply modules/imgproc/src/hough.cpp]: std::vector has clear behavior and implementation. std::stack is unknown thing. Let's keep std::vector.

File: modules/imgproc/include/opencv2/imgproc.hpp [hunk 3]
```diff
@@ -473,7 +473,8 @@ enum HoughModes {
     /** multi-scale variant of the classical Hough transform. The lines are encoded the same way as
     HOUGH_STANDARD. */
     HOUGH_MULTI_SCALE   = 2,
-    HOUGH_GRADIENT      = 3 //!< basically *21HT*, described in @cite Yuen90
+    HOUGH_GRADIENT      = 3, //!< basically *21HT*, described in @cite Yuen90
+    HOUGH_GRADIENT_ALT  = 4
```
[the remark this answers is not in the document]
Comment (vpisarev) [reply modules/imgproc/include/opencv2/imgproc.hpp]: done

File: modules/imgproc/src/hough.cpp [hunk 4]
```diff
@@ -1710,6 +1711,467 @@ static void HoughCirclesGradient(InputArray _image, OutputArray _circles,
...
+    double rw;
+    int weight;
+    uint64 mask;
+};
+
+static int circle_popcnt(uint64 mask)
+{
+    int count = 0;
+    for(int b = 0; b < 64; b++, mask >>= 1)
+        count += (mask & 1) != 0;
```
[the remark this answers is not in the document]
Comment (vpisarev) [reply modules/imgproc/src/hough.cpp]: I have not found simple inline function that takes uint64(_t) on input, just universal intrinsic that I do not need. I improved the implementation to use CV_POPCOUNT_U64 if possible. Someone (maybe me) in some subsequent patch can move it to some of core headers and rename cv::circle_popcnt() to cv::popcount() or cvPopCount64() or something like that.

File: modules/imgproc/src/hough.cpp [hunk 5]
```diff
@@ -1710,6 +1711,467 @@ static void HoughCirclesGradient(InputArray _image, OutputArray _circles,
...
+    HOUGH_CIRCLES_ALT_MAX_CLUSTERS = 10
+};
+
+static void HoughCirclesAlt( const Mat& img, std::vector<EstimatedCircle>& circles, double dp, double rdMinDist,
+                             double minRadius, double maxRadius, double cannyThreshold, double minCos2 )
+{
+    const int MIN_COUNT = 10;
+
+    if( maxRadius <= 0 ) maxRadius = std::min(img.cols, img.rows)*0.5;
+    if( minRadius > maxRadius ) std::swap(minRadius, maxRadius);
```
[the remark this answers is not in the document]
Comment (vpisarev) [reply modules/imgproc/src/hough.cpp]: done

File: modules/imgproc/src/hough.cpp [hunk 6]
```diff
@@ -1710,6 +1711,467 @@ static void HoughCirclesGradient(InputArray _image, OutputArray _circles,
...
+    maxRadius = std::max(maxRadius, 1.);
+    minRadius = std::max(minRadius, 1.);
+    minRadius = std::min(minRadius, maxRadius);
+    cannyThreshold = std::max(cannyThreshold, 1.);
+    dp = std::max(dp, 1.);
+
+    Mat Dx, Dy, edges;
+    Scharr(img, Dx, CV_16S, 1, 0);
+    Scharr(img, Dy, CV_16S, 0, 1);
+    Canny(Dx, Dy, edges, cannyThreshold/2, cannyThreshold, true);
```
[the remark this answers is not in the document]
Comment (vpisarev) [reply modules/imgproc/src/hough.cpp]: it's not preprocessing. Canny edge detector is the essential part of HoughCircles that takes grayscale image on input. HOUGH_GRADIENT method does the same.

File: modules/imgproc/src/hough.cpp [hunk 7]
```diff
@@ -1710,6 +1711,467 @@ static void HoughCirclesGradient(InputArray _image, OutputArray _circles,
...
+
+                float mag = std::sqrt((float)vx*vx+(float)vy*vy);
+                nz.push_back(Vec4f((float)p.x, (float)p.y, (float)vx, (float)vy));
+                CV_Assert(mdata[p.y*mstep + p.x] == 1);
+
+                int sx = cvRound(vx * 1024 / mag);
+                int sy = cvRound(vy * 1024 / mag);
+
+                int x0 = cvRound((p.x * idp) * 1024);
+                int y0 = cvRound((p.y * idp) * 1024);
```
[the remark this answers is not in the document]
Comment (vpisarev) [reply modules/imgproc/src/hough.cpp]: fixed

File: modules/imgproc/src/hough.cpp [hunk 8]
```diff
@@ -1710,6 +1711,467 @@ static void HoughCirclesGradient(InputArray _image, OutputArray _circles,
...
+                if( neighbors == 0 )
+                {
+                    if( backtrace_mode )
+                        nz.pop_back();
+                    backtrace_mode = true;
+                }
+                else
+                    backtrace_mode = false;
+            }
+            while(!stack.empty());
```
[the remark this answers is not in the document]
Comment (vpisarev) [reply modules/imgproc/src/hough.cpp]: done

File: modules/imgproc/src/hough.cpp [hunk 9]
```diff
@@ -1710,6 +1711,467 @@ static void HoughCirclesGradient(InputArray _image, OutputArray _circles,
...
+                if( (double)dv*dv < (double)minCos2*mag2*rij2 && i < nnz-1 ) continue;
+                float rij = std::sqrt(rij2);
+
+                CircleData& arc_j = arc[j];
+                double r_arc = arc_j.rw/arc_j.weight;
+                int di0 = 0;
+                int prev = prev_idx[j];
+                prev_idx[j] = i;
+
+                if( std::abs(rij - r_arc) < (r_arc + 80)*0.03 && prev+1 == i && !stop_marker )
```
[the remark this answers is not in the document]
Comment (vpisarev) [reply modules/imgproc/src/hough.cpp]: fixed

File: modules/imgproc/test/test_houghcircles.cpp [hunk 10]
```diff
@@ -181,102 +181,139 @@ INSTANTIATE_TEST_CASE_P(ImgProc, HoughCirclesTestFixture, testing::Combine(
...
 
     GaussianBlur(src, src, Size(9, 9), 2, 2);
 
-    vector<Vec3f> circles;
-    vector<Vec4f> circles4f;
-    HoughCircles(src, circles, CV_HOUGH_GRADIENT, dp, minDist, edgeThreshold, accumThreshold, minRadius, maxRadius);
-    HoughCircles(src, circles4f, CV_HOUGH_GRADIENT, dp, minDist, edgeThreshold, accumThreshold, minRadius, maxRadius);
+    for( int k = 1; k < 2; k++ )
+    {
+        int method = k == 0 ? HOUGH_GRADIENT : HOUGH_GRADIENT_ALT;
```
[the remark this answers is not in the document]
Comment (vpisarev) [reply modules/imgproc/test/test_houghcircles.cpp]: done

**Comment by vpisarev** [comment]
@alalek, the original issue complains about HoughCircles quality on some particular images, but the problem is much deeper. For a long time I knew that cv::HoughCircles() does not work well. But I was not sure about the effect of the latest modifications. So, for confirmation I took several revisions of OpenCV and ran them on a dataset of ~100 images (mostly from here: https://github.com/AlanLuSun/Circle-detection, but also my little collection of circles images and photos from Internet). The results are really bad. The existing implementation is very advanced in terms of speed optimization efforts, but it's fundamentally wrong. The newly proposed algorithm, on which I spent the whole week, uses the similar first phase with 2D accumulator update, but the center selection and the radius estimation are completely different. It's not compatible with the existing implementation. It treats minRadius a bit differently. It treats param2 completely differently, because accumThreshold is very difficult to set to some "universally good" value, you need it to vary from image to image and too bad if you have circles of different size - there is no good accumThreshold value in such case. So I decided to make it a separate algorithm without having to keep compatibility with the previous version. In OpenCV 5.0, perhaps, HOUGH_GRADIENT can be removed completely.

**Comment by catree** [comment]
If `HOUGH_GRADIENT_ALT` performs significantly better than `HOUGH_GRADIENT` with comparable speed performance, add somewhere in the documentation that `HOUGH_GRADIENT_ALT` method should be preferred?
Otherwise, a note that `HOUGH_GRADIENT_ALT` is more accurate while `HOUGH_GRADIENT` is faster?

Also, it would be great to have some numbers to evaluate the performance of the new method on a ground truth dataset.

**Comment by vpisarev** [comment]
@catree, thanks! More detailed test that evaluates performance over a real dataset is being developed.

**Comment by ccl-private** [comment]
@vpisarev Thanks, your implementation is much better. Would you introduce your algorithm in detail or is there a blog to introduce it? I really want to learn about your algorithm.

**Comment by feature-engineer** [comment]
@vpisarev When will a test evaluating the performance of the hough circles inplementations be published?

**Comment by nadavv** [comment]
Hi,
As I am new here, could you please explain what it would take for this version to be available in python?
Muchas gracias

**Pull request merged**